apis/condition_set_impl_test.go

```diff
@@ -840,3 +840,94 @@ func TestTerminalInitialization(t *testing.T) {
 		t.Errorf("GetCondition(Bar) = %s, wanted %s", got, want)
 	}
 }
+
+func TestRemoveNonTerminalConditions(t *testing.T) {
+	set := NewLivingConditionSet("Foo")
+	status := &TestStatus{}
+
+	manager := set.Manage(status)
+	manager.MarkTrue("Foo")
+	manager.MarkTrue("Bar")
+
+	if got, want := len(status.c), 3; got != want {
+		t.Errorf("Marking true() = %v, wanted %v", got, want)
+	}
+
+	if !manager.IsHappy() {
+		t.Error("IsHappy() = false, wanted true")
+	}
+
+	manager.RemoveCondition("Bar")
+	if got, want := len(status.c), 2; got != want {
+		t.Errorf("Marking true() = %v, wanted %v", got, want)
+	}
+
+	if !manager.IsHappy() {
+		t.Error("IsHappy() = false, wanted true")
+	}
+
+}
+func testCondition(t *testing.T, got *Condition, want corev1.ConditionStatus) {
+	if got == nil {
+		t.Errorf("GetCondition = nil, wanted %s", want)
+	}
+	if status := got.Status; status != want {
+		t.Errorf("GetCondition = %s, wanted %s", status, want)
+	}
+
+}
+func TestRemoveTerminalConditions(t *testing.T) {
+	set := NewLivingConditionSet("Foo", "Bar")
```

Comment:
can you also make a test that just has a condition set of "Foo" and then remove "Foo". So then happy has no dependents....

Reply:
@n3wscott what is the expected result here? When all dependent conditions are removed, does happy condition also need to be removed(like the initial case where there are no conditions)?

Reply:
A resource needs to have a happy condition. Ready in most cases. There are resources that do not have dependents, and using this method you could simulate what k8s resources do and when you are fully ready, you delete all deps and Ready remains true, and then when a depdent changes, you add it back to the deps list and set that dep to the new state (unknown or false). This would result in ready unknown or false and just the unknown or false dep being displayed in status.

Reply:
FWIW, feedback from client UX indicated that in the typical case, customers expect to see a consistent set of Conditions, even if they go "green" very quickly, rather than having some Conditions disappear after they are ready.

That's not to say that this functionality shouldn't exist, just a warning on UX.

Reply:
Point taken, this feature is for FYI conditions that do not apply to modes that the resource can transition into. And in that case, the condition as it is set does not make sense for the updated mode. And this give hooks to allow the deletion of that condition so the user does not see it and get confused.